llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -11660,6 +11660,24 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
     }
   }
 
+  // select m, sub(X, C), X --> sub (X, and(C, m))
+  if (N1.getOpcode() == ISD::SUB && N1.getOperand(0) == N2 && N1->hasOneUse() &&
+      DAG.isConstantIntBuildVectorOrConstantInt(N1.getOperand(1)) &&
+      N0.getScalarValueSizeInBits() == N1.getScalarValueSizeInBits()) {
```

Comment:
We need a DAG.ComputeNumSignBits(N0) == N0.getScalarValueSizeInBits() check as well

Reply:
Shouldn't `sext` guarantee that?